### ISSUE TITLE
Add ability to set text in small caps?

### ISSUE THREAD

**Issue by nickidlugash** [mapbox/mapbox-gl-js#5547]
All uppercase letters is a common styling technique for text in latin, cyrillic, and greek scripts. We use `text-transform: uppercase` to set this styling. This is not ideal for multilingual maps, because text using scripts that distinguish upper and lowercase letters (latin, cyrillic, greek) appear visually larger than text in other scripts.
	
- Mixed-script labels are not visually balanced between the two scripts:

  ![screen shot 2017-10-27 at 12 43 40 pm](https://user-images.githubusercontent.com/5607844/32124441-c4f83196-bb35-11e7-9ab8-ad6b6cc09f54.png)
  *E.g. these "Suburb" labels in Hong Kong, where the English more prominent*

- Labels in different regions are visually different sizes, which is problematic for maintaining a uniform label hierarchy across different regions:

  ![screen shot 2017-10-27 at 12 51 34 pm](https://user-images.githubusercontent.com/5607844/32124466-de758f38-bb35-11e7-97a9-b51fb97d7904.png)
  ![screen shot 2017-10-27 at 12 47 35 pm](https://user-images.githubusercontent.com/5607844/32124465-de6a66ee-bb35-11e7-85e0-11ade9ac3d2c.png)
  *E.g. these "Suburb" labels in London vs Tel Aviv* 

I think it would be interesting to explore implementing some form of display for "small capitals" (glyphs that mimic uppercase letters, but sized appropriately to be incorporated within upper-and-lowercase text). Although the main use case for small caps is mostly irrelevant here (especially since we can't currently style text differently within a single label), perhaps it's worth implementing for the use case outlined above?

Here's an inital braindump. Would be curious to hear others' thoughts on whether this seems worthwhile, and if so, how we might want to proceed.

###  True small caps

True small caps are glyphs specifically designed for this purpose, with appropriate stroke weights and proportions.

Typographically, displaying these glyphs would be ideal, but:  

- These don't exist for many font families, especially san serifs (which are more common on web maps).
- I think we might need to use an [OpenType text shaping engine (e.g. Harfbuzz)](https://github.com/mapbox/mapbox-gl-native/issues/7528) to be able to make use of these glyphs (but perhaps we don't? @chrisloer? )

### "Fake" small caps

"Fake" small caps are just uppercase letters with a reduced text size (usually such that the new cap height is slightly larger than the old x-height).

I think most renderers that support "small caps" fall back to this approach if the specified font does not contain true small cap glyphs.

Usually fake small caps are a big typographic no-no, largely because just shrinking the uppercase letters means that their strokes become too thin, relative to the other characters displayed in that font at that size. However:

 - Since for the most part our usage would only ever pair small cap letters in the same label with characters in scripts other than latin, cyrillic, or greek, perhaps the discrepency in text weight won't be so noticeable (especially in the cases where we're not even using the same font for different scripts).
 - There might be ways to programmatically increase the apparent text weight of fake small caps (although it seems like getting consistently good results would be difficult).

This would mean that for some labels, we'd be technically rendering part of the text in one text size and the other part in another text size (for some multi-script labels). Is that possible? 

### Spec

The CSS equivalent is `font-variant-caps: all-small-caps` or the shorthand [`font-variant: all-small-caps`](https://developer.mozilla.org/en-US/docs/Web/CSS/font-variant), which use the OpenType [feature tags](https://www.microsoft.com/typography/otspec/features_pt.htm) for supporting small caps: `smcp` (small  capitals) `c2sc` (capital to small caps). Perhaps we could support a similar property?

### Misc

Not directly related, but here's another ticket regarding better language support for uppercase styling: [text-tranform is not locale-aware](https://github.com/mapbox/mapbox-gl-js/issues/3999)

/cc @ChrisLoer @ansis @kkaefer @jfirebaugh @1ec5

**Comment by ChrisLoer** [comment]
> I think we might need to use an OpenType text shaping engine (e.g. Harfbuzz) to be able to make use of these glyphs (but perhaps we don't? @ChrisLoer?)

We can't do the normal thing of asking Harfbuzz to draw in small caps, but as long as the font has the glyph and we transliterate to the "small cap" code point (e.g. `latin letter small capital a`), the existing rendering engine could show the glyph. I just looked, though, and I'm surprised to learn that Unicode doesn't yet contain a `latin letter small capital q` or a `latin letter small capital x`!

As for doing "fake small caps" in our rendering code by changing the font-size on a glyph-by-glyph basis, yeah it should be possible although it might get a little complicated.

**Comment by khaledhosny** [comment]
I don’t think Unicode small caps are supposed to be used for typographic smallcaps and are very likely to give inferior results if used as such since they are encoded for use in phonetic notation and are to be used in a different context.

**Comment by 1ec5** [comment]
Agreed, those characters are meant for phonetic and occasional mathematical usage. Besides missing parts of even the basic English alphabet, the kerning is not what you’d expect for small caps.

> The CSS equivalent is `font-variant-caps: all-small-caps` or the shorthand [`font-variant: all-small-caps`](https://developer.mozilla.org/en-US/docs/Web/CSS/font-variant), which use the OpenType [feature tags](https://www.microsoft.com/typography/otspec/features_pt.htm) for supporting small caps: `smcp` (small  capitals) `c2sc` (capital to small caps). Perhaps we could support a similar property?

For what it’s worth, Core Text (mapbox/mapbox-gl-native#7528) has a `kUpperCaseSmallCapsSelector` option for small caps, plus all the other TrueType features listed in [this reference](https://developer.apple.com/fonts/TrueType-Reference-Manual/RM09/AppendixF.html).

**Comment by nickidlugash** [comment]
> For what it’s worth, Core Text (mapbox/mapbox-gl-native#7528) has a kUpperCaseSmallCapsSelector option for small caps, plus all the other TrueType features listed in this reference (example usage).

My understanding is that moving over to HarfBuzz is a not going to happen in the near future, given the huge lift required. If possible, it would be great to evaluate the feasibility of a small caps implementation before this point.

**Comment by kkaefer** [comment]
@nickidlugash could you provide a few mockups of how you imagine the text size to compare across different scripts? One thing we could do is to modify the glyph size for various scripts when *generating* SDFs so that latin and non-latin glyphs have a better size relation.

**Comment by nickidlugash** [comment]
@kkaefer Upper-and-lowercase text in latin, cyrillic, and greek scripts look appropriately sized in relation to text in other scripts – It's only all-uppercase text where the relationship is unbalanced (since all-uppercase glyphs are not designed to be typeset in a text block with other glyphs). If we were to modify glyph size when generating SDFs, I think that we would still need to generate glyphs in the original size, and only generate smaller glyphs for use with `text-transform: uppercase`. Would something like that be possible/reasonable?

**Comment by nickidlugash** [comment]
@kkaefer any thoughts on my above comment? Is there anything I can clarify better regarding design intent here?
